cpu/sam0_common/periph/eth.c

```diff
@@ -0,0 +1,316 @@
+           of the next frame if available */
+        if (max_len) {
+            /* If buffer available, copy data into it */
+            if (data)  {
+                memcpy(&data[rxlen], rx_buf[idx], len);
+            }
+            /* Tell the GMAC IP that we don't need this frame anymore  */
+            rx_curr->address &= ~DESC_RX_ADDR_OWNSHP;
+        }
+
+        rxlen += len;
+
+        if (rx_curr->status & DESC_RX_STATUS_END_FRAME) {
+            /* We reach the end of frame, leave the loop */
+            break;
+        }
+        /* Prepare next buffer */
+        idx = (idx + 1) % ETH_RX_BUFFER_COUNT;
+        rx_curr = &rx_desc[idx];
+
+   }
+   /* restore the previous index if packets were not released */
+   if (!max_len) {
+       rx_curr = &rx_desc[rx_idx];
+   }
+   /* Prepare next buffer if packets were released */
+   else {
+       rx_idx = (idx+1) % ETH_RX_BUFFER_COUNT;
+       rx_curr = &rx_desc[rx_idx];
+   }
+
+    return rxlen;
+}
+
+int sam0_eth_receive_blocking(char *data, unsigned max_len)
+{
+    return _try_receive(data, max_len, 1);
+}
+
+static void _enable_clock(void)
+{
+    /* Enable GMAC clocks */
+    MCLK->AHBMASK.reg |= MCLK_AHBMASK_GMAC;
+    MCLK->APBCMASK.reg |= MCLK_APBCMASK_GMAC;
+}
+
+int sam0_eth_init(void)
+{
+    /* Enable clocks */
+    _enable_clock();
+    /* Initialize GPIOs */
+    gpio_init_mux(sam_gmac_config[0].refclk, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].txen, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].txd0, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].txd1, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].crsdv, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].rxd0, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].rxd1, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].rxer, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].mdc, GPIO_MUX_L);
+    gpio_init_mux(sam_gmac_config[0].mdio, GPIO_MUX_L);
+    /* PHY reset */
+    gpio_init(sam_gmac_config[0].rst_pin, GPIO_OUT);
+    gpio_clear(sam_gmac_config[0].rst_pin);
+
+    /* reset buffers */
+    memset(rx_buf, 0, sizeof(rx_buf));
+    memset(tx_buf, 0, sizeof(tx_buf));
+    memset(rx_desc, 0, sizeof(rx_desc));
+    memset(tx_desc, 0, sizeof(tx_desc));
+
+    /* Enable PHY */
+    gpio_set(sam_gmac_config[0].rst_pin);
+
+    /* Initialize buffers descriptor */
+    _init_desc_buf();
+    /* Disable RX and TX */
+    GMAC->NCR.reg &= ~(GMAC_NCR_RXEN | GMAC_NCR_TXEN);
+    /* Enable Port Management */
+    GMAC->NCR.reg |= GMAC_NCR_MPE;
+    /* TODO: Implements MII, default to RMII */
+    GMAC->UR.reg = 0;
+    GMAC->IDR.reg = 0xFFFFFFFF;
+    GMAC->RSR.reg = GMAC_RSR_HNO | GMAC_RSR_RXOVR | GMAC_RSR_REC | GMAC_RSR_BNA;
+    /* Enable needed interrupts */
+    GMAC->IER.reg = GMAC_IER_HRESP | GMAC_IER_ROVR | GMAC_IER_TCOMP |
+                    GMAC_IER_TFC | GMAC_IER_RLEX | GMAC_IER_TUR |
+                    GMAC_IER_RXUBR | GMAC_IER_RCOMP;
```

Comment:
Wouldn't `GMAC_IER_RCOMP` be enough? That's the only event you are handling anyway.

Reply:
Done